### ISSUE TITLE
question: how are tests run using this action?

### ISSUE THREAD

**Issue by jkittner** [circuitpython/circuitpython-unified-build-ci#6]
Hi 👋 
I was seeing this action being used, instead of previous actions, across multiple repositories, where some of them also have tests. I had a look at [build/action.yml](https://github.com/adafruit/workflows-circuitpython-libs/blob/main/build/action.yml) but did not see any step that is running the tests?

I think it would be important to also run all tests in CI if they are available to avoid regressions and have confidence the package is actually working as expected.

**Comment by tekktrik** [comment]
Currently, tests aren't run using any of the actions listed.  I think some have not been run for a while which doesn't help.  I think the best way to handle this is to have an action here that runs the tests if a `tests` folder is detected.  Is that something you're interested in helping with?

**Comment by 2bndy5** [comment]
> I think the best way to handle this is to have an action here that runs the tests if a tests folder is detected

This can be done in 1 workflow using multiple jobs in the called workflow:
```yml
jobs:
  get-lib-info:
    runs-on: ubuntu-latest
    outputs:
      has-tests: ${{ steps.lib-info.outputs.has-tests }}
    steps:
      - name: get repo info
        id: lib-info
        run: |
          has_tests=false
          if [[ -d 'tests' ]]; then
              has_tests=true
          fi
          echo "has-tests=$has_tests" >> $GITHUB_OUTPUT

  run-tests:
    needs: [get-lib-info]
    # output vars are always strings, not booleans
    if: needs.get-lib-info.outputs.has-tests == 'true'
    steps:
      - run: coverage -m run pytest
```

<details><summary>somewhat off-topic comment</summary>

This tactic could also be used to consolidate the release-pypi workflow and the release-gh workflow with some extra bash script:
```yml
        run: |
          deploy_pypi=false
          if [[ -f 'setup.py' || -f 'pyproject.toml' ]]; then
              deploy_pypi=true
          fi
          echo "deploy-pypi=$deploy_pypi" >> $GITHUB_OUTPUT
```
Although, I think you can just check if pypi credentials are provided (as optional inputs in this scenario).
```yml
on:
  workflow_call:
    inputs:
      pypi-username:
        required: false
        default: '__token__'
        type: string
    secrets:
      pypi-password:
        required: false
        default: '?!'

jobs:
  release-pypi:
    if: secrets.pypi-password != '?!'
    # ...
```

</details>

**Comment by 2bndy5** [comment]
Oops! I just realized these workflows are meant to be composite actions and not [reusable workflows](https://docs.github.com/en/actions/using-workflows/reusing-workflows). While there's only a few subtle differences, my above suggestion (excluding the off-topic part) could still apply. @tekktrik please disregard [my other comment about using secrets](https://github.com/adafruit/workflows-circuitpython-libs/pull/11#issuecomment-1401433210); I'm so sorry.

**Comment by tekktrik** [comment]
Fixed by #18
